### ISSUE TITLE
Programmatically change size (specifically to implement hiding a space when it's under certain sizes)

### ISSUE THREAD

**Issue by Slapbox** [aeagle/react-spaces#36]
**Is your feature request related to a problem? Please describe.**
I'd like to implement a feature to hide a space if it's reduced to a certain size. Programmatically setting state and passing that state to the component initially works, but can get out of sync with the pass props. I assume this is not surprising or intended to be possible at present?

**Describe the solution you'd like**
It would be great to be able programmatically set a size. A key place this might be used is in `dragStart` and `dragEnd` events https://github.com/aeagle/react-spaces/issues/32

```js
render() {
    const { size } = this.state;
    return (
      <Space.ViewPort>
        <Space.LeftResizable size={size} />>
        <Space.Fill />
      </Space.ViewPort>
    );
  }
```

**Describe alternatives you've considered**
A CSS and plain JavaScript eventListener-based approach, but meh.

**Comment by aeagle** [comment]
@Slapbox Could you give me a little more info on the use case for this functionality? Is it because you want certain spaces to disappear when the overall viewport is below a certain size, i.e. a sidebar that should not show when the window is too small?

**Comment by Slapbox** [comment]
@aeagle that's one way I'd planned to use it, although that particular use case could be achieved prety cleanly with media-queries alone.

Ideally it would be great to have more low-level control over spaces though. In particular, in my example above, the space I'd like to gain some more programmatic control over is `Space.LeftResizable` and yes, as something of a sidebar.

One more involved use in my particular application is to programmatically hide a particular space when some other element (inside a different space) is focused. This could presumably already be achieved by passing `className`s to the `Space.LeftResizable`, but we're in the process of moving everything away from CSS to [styled-components](https://github.com/styled-components/styled-components), so that approach it's a bit escape-hatchy.

**Comment by Slapbox** [comment]
I'm thinking you fixed the potential for getting out of sync here by setting `adjustedSize: 0` ?

https://github.com/aeagle/react-spaces/commit/72e48ccd6404df358b234c9f113ce6eeb9665654#diff-0047d8a24a58c9eefedb8ef09a5b3996R30

**Comment by Slapbox** [comment]
Specifically to recreate the "out-of-sync" issue:

1. Create a `Resizable` of a particular `size` that's managed via some custom component that wraps `Resizable`
2. Adjust the size with mouse drag
3. Set the `size` to a value identical to the previously passed `size` value

So technically this can be worked around in the present pretty easily by simply not using an identical `size` to your original.

The particular spot that the prop value comparison is made: https://github.com/aeagle/react-spaces/blob/72e48ccd6404df358b234c9f113ce6eeb9665654/react-spaces/src/Resizable.tsx#L31-L32

**Proposed (Vague) Solution:**

Resizable could store its real size in `state` then use `static getDerivedStateFromProps` (or if there's a hook equivalent) to reconcile incoming props with the current state so that resizing with a mouse or touch event wouldn't result in the component's actual size differing from its `adjustedSize` - if I'm understanding what's going on correctly.

**Comment by aeagle** [comment]
@Slapbox If I'm getting this correctly, the out-of-sync issue you are mentioned is something along these lines:

```
const Outer = () => {
  const [size, setSize] = useState(200);
  return (
    <Spaces.ViewPort>
      <Spaces.LeftResizable size={size} />
    </Spaces.ViewPort>
  );
};
```
After resizing the left space the state held by the outer component no longer reflects the actual size of the left space.

Internally a resizable space has a size (passed in via props) and a resize delta (the difference applied) after a resize action to determine it's actual size.

I think that feeding back the new size after a resize operation may be the solution here as you suggested. I'm in process of implementing the resize events on #32. 

If the resize end event fed back the final new size of the space, you could you it to update the size held at the outer component (which would therefore reset the resize delta to 0 within the space). 

Something like:

```
const Outer = () => {
  const [size, setSize] = useState(200);
  return (
    <Spaces.ViewPort>
      <Spaces.LeftResizable size={size} onResizeEnd={(newSize} => setSize(newSize)} />
    </Spaces.ViewPort>
  );
};
```

**Comment by Slapbox** [comment]
Yes that's pretty much what I'm thinking and looking to do.

I think what we discussed is probably the simplest solution, and I'm really not too familiar with hooks or the underlying code of `React-Spaces`, so I don't know if it's helpful at all, but I figured I should mention the existence of `getSnapshotBeforeUpdate`.

It seems even experienced React developers forget `getSnapshotBeforeUpdate` exists. I've personally never had occasion to use it in my hundreds and thousands of hours of React development, but I know at least one time where it proved to be a godsend for a problem on another project.

**Comment by Slapbox** [comment]
I'm going to close this because the API now provides the tools for someone to implement this on their own. Of course, if you feel the conditional hiding is within the scope of the project, by all means re-open.

Thanks again for your great work!

**Comment by Slapbox** [comment]
@aeagle it seems that one bug remains - specifically that after setting a size, lets say `size={100}`, if you then resize with the mouse, passing the `size={100}` will have no effect. Passing instead `size={100 + rand}` will work though. I'm not exactly sure _how_ this is the case since you're passing `props.size` directly to the `return`ed `div` here https://github.com/aeagle/react-spaces/blob/master/react-spaces/src/Resizable.tsx#L82-L84

I'm guessing at some higher level in the component tree there's something I'm overlooking.

Due to a bug in debugging in Electron, currently I can't inspect anything that takes place as a result of a mouse or keyboard event, making it extremely difficult for me to study the issue better. Somewhere though I think `size` needs to be stored in state, and I think currently it's handled solely with props.

This isn't a great example since I don't really understand how Spaces is doing its thing, but to at illustrate what I mean:

```js

//  Assume resizable used as a left-hand sidebar
class Resizable extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      size: props.size,
    }
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const { size: nextSize } = nextProps;
    return { size: nextSize } //  Handles changes that come from prop changes
  }

  onResize = evt => {
    this.setState({ size: evt.pageX })  //  Handles changes that come from user events
  }

  render() {
    return (
      <div 
        style={{ width: props.width, height: props.height }}
        className={`spaces-resize-handle ${props.type}`} 
        onMouseDown={startResize}
        onTouchStart={startTouchResize} />
    )
  }
}
```

[Or it might be easier to do this with memoization](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization) somewhere else.

**Comment by Slapbox** [comment]
@aeagle this isn't the best demo, but hopefully it demonstrates the last remaining issue I'm seeing well enough. https://recordit.co/2wczCiDu3v

**Explanation:**

The first time the space is collapsed down, I re-open it by simply clicking the resizer, programmatically setting the width to `250px`

The second time I click and drag as if to resize the space. In this case the size I set for the space (`250px`) gets added to the `evt.nativeEvent.pageX`. So:

`finalValue = 250 + evt.nativeEvent.pageX`

If a `minimumSize` is passed to a space, it seems that the resulting value is calculated as so instead:

`finalValue = 250 + minimumSize + evt.nativeEvent.pageX`

My code that produces the issue looks like this:

```js
const COLLAPSED_SIZE = 20;

class SomeComponent extends React.Component {
  onResizeStart = () => {
    const { size, minimumSize } = this.state;
    if (size !== COLLAPSED_SIZE) {
      this.setState({ isCurrentlyResizing: true });
    } else {
      this.blockResizeEnd = true;
      this.setState({ size: minimumSize });
    }
  };

  onResizeEnd = width => {
    if (this.blockResizeEnd) {
      this.blockResizeEnd = false;
      return;
    }
    //  ... the rest of the function's usual logic
  }

  render() {
    //  ... the render function
  }
}
```

**Comment by aeagle** [comment]
@Slapbox thanks for the screen recording and the information.

To give an insight on how a resizable space is sized, it is always the result of size you passed in +- the difference of pixels a mouse resize operation has made to the initial size. This is to allow percentage sized spaces to resized correctly (effectively resulting with a `width: calc(25% + 200px)` in the markup or `width: calc({props.size} + {state.adjustedDelta})`. The adjusted delta in fact already stored in state as you suggested.

The props.width and props.height you referred to in the `Resizable` component is in fact the size of the resize handle (the thing you click on to start a resize) and nothing to do with the size of the space.

I'll have a look into your examples shortly. My immediate thought into this is that changing a spaces size whilst in the middle of a resize operation would produce undesired effects and put the state out-of-sync.

**Comment by Slapbox** [comment]
Thanks for the explanation @aeagle!

Perhaps the easiest way to handle it would be to provide a callback a user can call to tell the space that the next `onResizeEnd` should be ignored because it was already handled by user-provided logic? That _seems_ a little hacky, but at the same time seems probably the simplest way to implement it, and quite flexible. You'll know better than me of course.

My method would require a user to give the `Space` a ref so a user could do:

```js
onResizeStart = () => {
    const { size, minimumSize } = this.state;
    if (size !== COLLAPSED_SIZE) {
      this.setState({ isCurrentlyResizing: true });
    } else {
      this.myReactSpace.current.ignoreNextResizeEnd(); // Space would not fire the next onResizeEnd callback
      this.setState({ size: minimumSize });
    }
  };
```

**Comment by aeagle** [comment]
Hi @Slapbox. Apologies for the delay on this. I have been reworking the state in this library to find a better way of avoiding out-of-sync issues you and others have been experiencing. I have creating an alpha package `v0.1.20-alpha.3` with these modifications. I wonder if you could try this out and see if you experience the same problems as before.

**Comment by Slapbox** [comment]
Hey @aeagle - no need for apologies, you rock. Thanks for the update!

Unfortunately `alpha.3` didn't resolve [this issue from this thread](https://github.com/aeagle/react-spaces/issues/36#issuecomment-545166275) - since I know there's a few distinct issues in this thread, nor did it resolve #47. Let me know any time you could use some testing and I'll be happy to oblige.

**Comment by Slapbox** [comment]
Absolutely. I'll ping you when I've got something.

**Comment by Slapbox** [comment]
As I was creating this repro it occurred to me that this is kind of in a gray area between _bug_ and _unanticipated use case_.

https://codesandbox.io/s/lucid-bose-ws6rq

**What would I expect to happen instead**

Upon resizing to the default size in `onResizeStart`, the space becomes aware of its new starting width and increases the space's size **_to the size_** the user dragged to, rather than _**by the amount**_ the user dragged.

**Comment by Slapbox** [comment]
It looks like there's another related issue. If I previously set the space size in pixels and then pass a value of `100%` I get `width: calc(100% + 296px);` instead of `width: calc(100%);` or just `width: 100%;`

**Comment by Slapbox** [comment]
Another related issue (not sure how to recreate) but I've had a few times where my `<LeftResizable />` and my `<Fill />` elements have both gotten identical values.

```css
#LeftResizable {
  width: calc(427px + 296px);
}

#Filll {
  left: calc(427px + 296px);
```

The best theory I have right now is that the `LeftResizable` was `296px` wide, and then when I tried to set it to the full with of the window (`427px`) it somehow ends up adding the existing `size` and the new one, rather than replacing the existing `size`.

**Comment by Slapbox** [comment]
I think I understand I understand the issue better now (or I could be totally wrong.) I'm not versed in Typescript or I'd take a crack at implementation. I don't know if this would fix minor related issues like being able to drag a space past its maximum size this way, or if my idea is incomplete.

https://github.com/aeagle/react-spaces/blob/master/react-spaces/src/Space.tsx#L176-L183

This covers all cases I can think of except this particular one:

1. Defining the space size in the parent like `<LeftResizable size={this.state.size} />`
2. Programmatically changing the size of a space upon initiating a resize, like:

```js
  onResizeStart = () => {
    this.setState({ size: 300 });
  }
```

The use case for doing this is to hide a space, but allow it to be shown again by clicking. Unfortunately a user can continue the drag operation after the click.

I think the logic being used here in `onResizeEnd` to get element bounds might be better run outside of `onResizeEnd` so we can pass the dimensions to `onResize` replacing the `originalX` and the `originalY` for the final run.

https://github.com/aeagle/react-spaces/blob/master/react-spaces/src/Space.tsx#L55
https://github.com/aeagle/react-spaces/blob/master/react-spaces/src/Space.tsx#L163

Here's a fragment of code demonstrating my proposal. I'm not sure about some typescript syntax so I used plain JS to avoid any confusion. I don't know the exact math that's needed there, maybe it's more complicated than I think?

https://gist.github.com/Slapbox/2ec285aeaac5c9cad8131097e86177cd

**Comment by aeagle** [comment]
@Slapbox - Thanks for the extra explanation. From what I gather the "Unfortunately a user can continue the drag operation after the click" stands out to me. It occurs to me that a collapsed space should not be resizable at all and you are trying to find the behaviour you need by shortcutting the resize.

If you take a look at this modified code sandbox, I've slightly re-implemented to give the behaviour you have described - https://codesandbox.io/s/vigilant-haibt-7l6jj.

The main difference here is that I am actually using a Custom space (this gives you access to all the properties between different types of spaces) allowing you to switch the space between a resizable and non resizable space depending on it's collapsed state using the `resizable` property. In it's collapsed state I am simply binding to the onClick event of the space to expand it again.

This is my view seperates the concerns of resizing and collapsing. I've exposed the Custom space in `0.1.22-alpha.1`. Would be good to know what you think of this?

**Comment by Slapbox** [comment]
@aeagle that's very slick and works great! Much, _much_ simpler than the route I was headed down.

The only issue for our use case is that we rely on the handle as the element the user clicks to restore the space, but with `resizable={false}`, the handle is (sensibly) removed from the DOM. What do you think of an additional option/props to allow for showing the resize handle, even when the space is not `resizable`?

**Comment by aeagle** [comment]
@Slapbox I think in that case you might consider putting your own element in the collapsed space which is clickable. Another Space.Fill perhaps or an element absolutely positioned with left, top, right, bottom at 0. What do you think?

**Comment by Slapbox** [comment]
I considered it and it wouldn't be the end of world, but it wouldn't be trivial to get it working seamlessly either. A number of things would need to be reconsidered/redesigned on our end. We're do something custom styling of hidden spaces, and specifically applying `opacity: 0` to the content of a hidden space for a smooth transition, while leaving the `resizeHandle`. There's some additional concerns styling concerns. Like I said, not the end of the world, but it would end up being quite complicated solution compared with having access to the same resize handle that's there when it's resizable.

What do you think of applying this CSS to the resize handle when it's not needed instead of dropping it from the DOM?

```css
  visibility: hidden;
  pointer-events: none;
```

This way a user could override the default behavior without adding any complexity to the library. I'm not sure if there's some reason maybe that keeping it in the DOM would be difficult though?

A heavier-duty approach (perhaps needlessly heavy-duty) would be to allow users to pass in their own element to be used as a the resize handle and allow them the option to manage the visibility completely on their own.

**Comment by aeagle** [comment]
@Slapbox I'm not sure about adding markup for the resize handle where the space is not resizable and hiding it. I'm eager to keep the markup to a minimum and only where necessary hence the recent changes to move the styling to the head tag and remove the need for the outer and inner tags. 

I have an alternative option for you which might suit. In the same version you tested `0.1.22-alpha.1` you could stick with the resizable space but effectively cancel the resize when the space is collapsed by returning `false` from callback you pass to the resize start event on the space.

**Comment by Slapbox** [comment]
I understand your hesitancy. 

Your suggestion of returning `false` works about 95% as I'd envisioned, which to me makes it a beautifully simple solution.

**Comment by aeagle** [comment]
@Slapbox Glad that helped! I've released these changes on `0.1.22` now.
